<?xml version="1.0"?>
<robot name="box_robot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:include filename="$(find homewhere)/urdf/color.xacro" />
  <xacro:include filename="$(find homewhere)/urdf/color.gazebo.xacro" />
  <xacro:include filename="$(find homewhere)/urdf/plugin.xacro" />

  <!-- ================== -->
  <!-- 0) Properties      -->
  <!-- ================== -->

  <xacro:property name="base_mass" value="20.0" />
  <xacro:property name="base_size" value="0.25 0.25 0.25" />

  <xacro:property name="steer_mass" value="2.0" />
  <xacro:property name="steer_size" value="0.05 0.05 0.1" />

  <xacro:property name="wheel_mass" value="2.0" />
  <xacro:property name="wheel_radius" value="0.07" />
  <xacro:property name="wheel_length" value="0.04" />

  <link name="base_link"></link>
  <link name="base_footprint" />
  <joint name="base_joint" type="fixed">
    <parent link="base_footprint" />
    <child link="base_link" />
    <origin xyz="0 0 0" rpy="0 0 0" />
  </joint>

  <!-- ================== -->
  <!-- 1) Base Link       -->
  <!-- ================== -->

  <link name="body">
    <visual>
      <geometry>
        <box size="${base_size}" />
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <geometry>
        <box size="${base_size}" />
      </geometry>
    </collision>

    <inertial>
      <mass value="${base_mass}" />
      <origin xyz="0 0 0" />
      <inertia
        ixx="0.208" ixy="0.0" ixz="0.0"
        iyy="0.208" iyz="0.0"
        izz="0.208" />
    </inertial>
  </link>

  <joint name="base_to_body" type="fixed">
    <parent link="base_link" />
    <child link="body" />
    <origin xyz="0 0 0" />
  </joint>

  <!-- ============== -->
  <!-- 2) Wheel       -->
  <!-- ============== -->

  <xacro:macro name="steering_wheel" params="name x y">

    <!-- roll = 1.57 to rotate the wheel upright-->
    <link name="wheel_${name}">
      <visual>
        <geometry>
          <cylinder length="${wheel_length}" radius="${wheel_radius}" />
        </geometry>
        <origin xyz="0 0 0" rpy="1.57 0 0" />
        <xacro:if value="${name == 'front_right'}">
          <material name="red" />
        </xacro:if>
        <xacro:unless value="${name == 'front_right'}">
          <material name="grey" />
        </xacro:unless>
      </visual>

      <collision>
        <geometry>
          <cylinder length="${wheel_length}" radius="${wheel_radius}" />
        </geometry>
        <origin xyz="0 0 0" rpy="1.57 0 0" />
      </collision>

      <inertial>
        <mass value="${wheel_mass}" />
        <origin xyz="0 0 0" />
        <inertia
          ixx="0.0027" ixy="0.0" ixz="0.0"
          iyy="0.0049" iyz="0.0"
          izz="0.0027" />
      </inertial>
    </link>

    <link name="steer_${name}">
      <visual>
        <geometry>
          <box size="${steer_size}" />
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0" />
        <xacro:if value="${name == 'front_right'}">
          <material name="red" />
        </xacro:if>
        <xacro:unless value="${name == 'front_right'}">
          <material name="green" />
        </xacro:unless>
      </visual>

      <inertial>
        <mass value="${steer_mass}" />
        <origin xyz="0 0 0" />
        <inertia
          ixx="0" ixy="0" ixz="0"
          iyy="0" iyz="0"
          izz="0" />
      </inertial>
    </link>

    <!-- z = base_side/2 + leg/2 -->
    <!-- <joint name="steer_${name}_joint" type="revolute">
      <parent link="body" />
      <child link="steer_${name}" />
      <origin xyz="${x} ${y} -0.1" rpy="0 0 0" />
      <axis xyz="0 0 1" />
      <limit lower="${-pi/2}" upper="${pi/2}" effort="1000.0" velocity="0.1" />
      <safety_controller soft_lower_limit="${-pi/2 * 0.9}" soft_upper_limit="${pi/2 * 0.9}" k_position="10.0" k_velocity="1000.0" />
    </joint> -->
    
    <!-- z = base_side/2 + leg/2 -->
    <joint name="steer_${name}_joint" type="continuous">
      <parent link="body" />
      <child link="steer_${name}" />
      <origin xyz="${x} ${y} -0.175" rpy="0 0 0" />
      <axis xyz="0 0 1" />
      <!-- <limit lower="${-pi/2}" upper="${pi/2}" effort="1000.0" velocity="0.1" />
      <safety_controller soft_lower_limit="${-pi/2 * 0.9}" soft_upper_limit="${pi/2 * 0.9}" k_position="10.0" k_velocity="1000.0" /> -->
    </joint>

    <!-- z = -leg/2 -->
    <joint name="wheel_${name}_joint" type="continuous">
      <parent link="steer_${name}" />
      <child link="wheel_${name}" />
      <origin xyz="0 0 -0.05" />
      <axis xyz="0 1 0" />
    </joint>

  </xacro:macro>

  <!-- ============== -->
  <!-- 3) Lidar       -->
  <!-- ============== -->

  <link name="base_scan">
    <visual>
      <!-- <origin xyz="0 0 0" rpy="0 0 0" /> -->
      <geometry>
        <cylinder length="0.05" radius="0.02" />
      </geometry>
      <material name="red" />
    </visual>
  </link>

  <joint name="scan_joint" type="fixed">
    <parent link="base_link" />
    <child link="base_scan" />
    <origin xyz="0 0 0.15" rpy="0 0 0" />
  </joint>

  <xacro:steering_wheel name="front_left" x="0.1" y="0.1" />
  <xacro:steering_wheel name="front_right" x="0.1" y="-0.1" />
  <xacro:steering_wheel name="rear_left" x="-0.1" y="0.1" />
  <xacro:steering_wheel name="rear_right" x="-0.1" y="-0.1" />

</robot>